libp2p/protocols/pubsub/gossipsub/behavior.nim

```diff
@@ -105,9 +105,13 @@ proc handleGraft*(g: GossipSub,
 
       continue
 
+    # Check backingOff
+    # Ignore BackoffSlackTime here, since this only for outbound activity
```

Comment:
I don't really follow the reason we're subtracting BackoffSlackTime twice here..?

when we send a backoff time to someone, our clock starts ticking but theirs doesn't until they process the message - that means that if they follow the protocol reasonably, they will not send a graft until `backoffTime+roundrip` so there's already some "natural" slack built in - is there any harm in removing the slack time completely?

Reply:
Yes, but it could be them that sent the `backoff`
Let's say at the moment they sent the `backoff` they had 500ms roundtrip, we'll have 500ms more backoff than them.

But then, for some reason, the roundtrip improves to 100ms, and they graft us as early as they can. That would be 400ms too early for us.

The "natural" slack is jittery and not very big, that is the reason that the `BackoffSlackTime` exists:
- We make sure that we send outgoing graft 2 seconds after the time we're allowed, to avoid race condition on that side
- We make sure that we allow incoming graft 2 seconds before they're allowed, to avoid on the other side

It may be cleaner to store the backoff without the Slack time, and compute it each it time though
But for all the cases except this one, the backoff is used for outgoing and not incoming

Reply:
well, still: when we receive a prune, we store their graft time - when we send a prune, we store our graft time - it seems to me that the max of these two values should be used to decide when to send the next graft, no? or do we not receive any prune from them when we send a prune? can't remember.

Reply:
Yes, the peer pruning is sending his backoff, and the pruned peer doesn't have anything to say about it

Reply:
AFAIR,  pruning is send and forget and no ack is implied from the pruned peer.